### PR TITLE
Create Error State Interface

### PR DESCRIPTION
**Pull request by sawsa307**
Added error checking functions in transactionSyncer.go for syncer's error state get/set/reset.

### CODE REVIEW

File: pkg/neg/syncers/transaction.go [hunk 1]
```diff
@@ -93,6 +93,9 @@ type transactionSyncer struct {
 	enableEndpointSlices bool
 
 	logger klog.Logger
+
+	// grab syncLock first if any decision needs to be based on this value
+	inError bool
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:98]: Can you expand the comment to indicate what it means to be in error state?

File: pkg/neg/syncers/transaction.go [hunk 2]
```diff
@@ -184,6 +188,11 @@ func (s *transactionSyncer) syncInternal() error {
 }
 
 func (s *transactionSyncer) syncInternalImpl() error {
+	// TODO: enter degraded mode before reset and sync
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:191]: Can you expand the comment to indicate when do we enter the error state and why we will reset here? 

What was the finalized decision of when we run in degraded mode?

**Review by swetharepakula — COMMENTED** [review]
Couple of nits on the comments. Please update the commit message since we no longer have the error state interface.

File: pkg/neg/syncers/transaction.go (hunk 1 above)
Comment (sawsa307) [reply pkg/neg/syncers/transaction.go:103]: Update the comments.

File: pkg/neg/syncers/transaction.go (hunk 2 above)
Comment (sawsa307) [reply pkg/neg/syncers/transaction.go]: Updated the comment. We will run in degraded mode at the beginning of next sync, and after degraded mode procedures are done, we reset the error boolean.

File: pkg/neg/syncers/transaction.go [hunk 3]
```diff
@@ -93,6 +93,14 @@ type transactionSyncer struct {
 	enableEndpointSlices bool
 
 	logger klog.Logger
+
+	// inError indicates if the syncer is in any of 4 error scenarios
+	// 1. Endpoint counts from EPS is different from calculated endpoint list
+	// 2. EndpontSlice has missing or invalid data
+	// 3. Attach/Detach EP fails due to incorrect batch information
+	// 4. Endpoint count from EPS or calculated endpoint list is 0
+	// Need to grab syncLock first if any decision needs to be based on this value
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:102]: nit: lock is needed for writes as well.

**Review by swetharepakula — COMMENTED** [review]
One nit, otherwise this LGTM. Thanks David!

File: pkg/neg/syncers/transaction.go (hunk 3 above)
Comment (sawsa307) [reply pkg/neg/syncers/transaction.go]: Fixed. Thanks!

**Review by swetharepakula — COMMENTED** [review]
Thanks David!

/lgtm
/approve

**Pull request merged**